src/main/java/org/passay/PasswordGenerator.java

```diff
@@ -151,29 +172,151 @@ public UnicodeString generatePassword(final int length, final List<? extends Rul
+          fillChars = chars.difference(illegalChars, true);
+        } else {
+          fillChars = chars;
+        }
+      } else {
+        fillChars = new UnicodeString();
+      }
+    }
+    return fillChars;
   }
 
 
   /**
-   * Fills the supplied target with count random characters from source.
+   * Returns the list of character appenders used for password generation.
    *
-   * @param  source  of random characters.
-   * @param  count  number of random characters.
-   * @param  target  character sequence that will hold characters.
+   * @param  rules  to derive password characters
+   * @param  allowedChars  characters allowed
+   * @param  illegalChars  characters not allowed
+   * @param  rand  to randomize character selection
+   *
+   * @return  character appenders
    */
-  protected void fillRandomCharacters(final String source, final int count, final CharBuffer target)
+  protected List<CharacterAppender> getCharacterAppenders(
+    final List<? extends Rule> rules,
+    final UnicodeString allowedChars,
+    final UnicodeString illegalChars,
+    final Random rand)
   {
-    final int[] indexes = codePointIndexes(source);
-    for (int i = 0; i < count; i++) {
-      final String s = PassayUtils.toString(source.codePointAt(indexes[random.nextInt(indexes.length)]));
-      if (target.position() + s.length() <= target.limit()) {
-        target.append(s);
+    final List<CharacterAppender> appenders = new ArrayList<>();
+    for (Rule rule : rules) {
+      if (rule instanceof CharacterRule) {
+        appenders.add(new CharacterRuleAppender((CharacterRule) rule, allowedChars, illegalChars, rand));
+      } else if (rule instanceof CharacterCharacteristicsRule) {
+        appenders.add(
+          new CharacterCharacteristicsAppender((CharacterCharacteristicsRule) rule, allowedChars, illegalChars, rand));
       }
     }
+    return appenders;
+  }
+
+
+  /**
+   * Return a unicode string that contains the unique characters from all {@link CharacterRule} or
+   * {@link CharacterCharacteristicsRule} contained in the supplied list.
+   *
+   * @param  rules  to extract unique characters from
+   *
+   * @return  unicode string containing unique characters or empty string
+   */
+  private UnicodeString getCharacters(final List<? extends Rule> rules)
+  {
+    final List<CharacterRule> characterRules = new ArrayList<>();
+    for (Rule rule : rules) {
+      if (rule instanceof CharacterRule) {
+        characterRules.add((CharacterRule) rule);
+      } else if (rule instanceof CharacterCharacteristicsRule) {
+        characterRules.addAll(((CharacterCharacteristicsRule) rule).getRules());
+      }
+    }
+    UnicodeString chars = null;
+    if (!characterRules.isEmpty()) {
+      chars = new UnicodeString(characterRules.get(0).getValidCharacters());
+      for (int i = 1; i < characterRules.size(); i++) {
+        chars = chars.union(new UnicodeString(characterRules.get(i).getValidCharacters()), true);
```

Comment:
Given that this is not a credential or other sensitive data, do you need to clear arrays here?